### PR TITLE
Issues/66 rewrite

### LINKED ISSUE #66: Rewrite?
Instead of launching a new `WordPress_Radio_Taxonomy` for each taxonomy (which adds a bunch of repetitive hooks), merge into a single file and use conditional logic on the hooks to test if the taxonomy is one of the RB4T-enabled taxonomies.

### LINKED ISSUE #93: Remove meta boxes called many times on all admin pages
Hi Helga, 

I was looking for ways to speed up Admin performance and came across an issue that also plays in this plugin. 

The remove_meta_box function, this one:

class.WordPress_Radio_Taxonomy.php 
add_action in line 46
function remove_meta_box in line 76

This function is being called many times on all admin pages, where it is only needed on post type edit screens where there is a radio metabox, and only once for each radio taxonomy that post type is connected to.

In my case my taxonomies that are set to have radio buttons are connected to multiple post types. 
The remove_meta_box function is called for every posttype it is connected to, for each of these taxonomies, on every pages. This adds up quickly:

Specifically in my case I have 4 radio taxonomies. Each is connected to 4 custom post types. This results in 16 calls to remove_meta_box, where only one or two are needed. And even worse, these 16 calls happen on ALL admin pages, also on post types that have no radio buttons, and literally every admin page like Plugins, Media etc.

I use something like below to prevent this, in other scenario's where I know the post type in advance. I check for get_current_screen(), and let it only run if it is needed, checking if it is present in an array of posttypes. Would have to be adapted of course.

Also, more or less the same happens for add_meta_box function. 

I am hoping you are able to fix this. Thanks!

```
function remove_unwanted_metaboxes( ) {

	// The admin pages that don't have get_current_screen  won't have metaboxes 
	if ( function_exists( 'get_current_screen' ) ) {
		
		$postType = get_current_screen()->post_type;

		$postTypes = array(
			'page',
			'post'
		); //this should be dynamic of course

		if ( in_array( $postType, $postTypes ) ) {

			remove_meta_box( 'mymetaboxid', $postType, 'side' );

		}
	}
}
add_action( 'admin_head', 'remove_unwanted_metaboxes' );
```

### PR DESCRIPTION
**Pull request by helgatheviking**
Launch the taxonomy customizations only on the taxonomies that are enabled as radio. Closes #93. Closes #66.

cc/ @remcokalf  made a few tweaks to your PR. could you give it a test run?

### CODE REVIEW

**Comment by helgatheviking** [comment]
Would like to resolve #91 and ship it

**Comment by remcokalf** [comment]
Will test it now.

**Comment by remcokalf** [comment]
I don't see any changes from my PR. Did you not push your last commit? Or am I missing something?

**Comment by helgatheviking** [comment]
I forgot to push! fail!

**Comment by remcokalf** [comment]
All looks good except one thing that caught my eye: In the metabox, the checkboxes in the 'Most Used' tab are not converted to radio buttons. I think this was working before.

**Comment by helgatheviking** [comment]
Class fix 1 thing and break another. Wonder where I went wrong....

**Comment by helgatheviking** [comment]
Copy/paste fail! Used the wrong action hook name and callback. Should be good now. 🤞

**Comment by remcokalf** [comment]
Yeah I had been wondering how that hook was working, but it worked mysteriously well :-) through the deprecated add_meta_box() function.

Anyways, I tested again, all looks good.

Except one small issue:

On line 102 of class-wordpress-radio-taxonomy.php you call `add_meta_box()` with the second parameter set to:
`$this->tax_obj->labels->singular_name`. This should be `$this->tax_obj->labels->name` I think, as this is what the header of the metabox is originally.

I noticed because I have quite some custom taxonomies with a singular name different than the name.
And also because I have another plugin Tabify Edit Screen, that manipulates metaboxes onto tabs in post edit screens. In the settings of that plugin, all removed radio metaboxes now have different name than the replaced metaboxes, so they show up twice.

If that is solved, I think it's good to go. Don't forget the version bump, that is still 2.3.3.

**Comment by remcokalf** [comment]
I might have one more thing, currently testing. This Tabify Edit Screen plugin is consistently not working right with the radio metaboxes now. Maybe a hook order issue.

**Comment by helgatheviking** [comment]
Interestingly, it wasn't working at ALL. :) It's the filtering the `wp_terms_checklist_args` that gets some of the radios. To convert the popular checklist as well is why I needed to manually replace the entire metabox.

As for the metabox label, yes it does look like it is `$taxonomy->labels->name` in core:
https://github.com/WordPress/WordPress/blob/master/wp-admin/includes/meta-boxes.php#L1479

Do you think this should be the same for the "No term" option? Currently that is
```
$no_term = sprintf( __( 'No %s', 'radio-buttons-for-taxonomies' ), $this->tax_obj->labels->singular_name );
```

(would need to check how it is in the block editor, but I think it's the same)

Version bump will be handled by a build script I'll run before deploying to .org..... so all good there!

I'm not familiar with the Tabify Edit Screen plugin at all. And full-disclosure, like a lot of free plugins at .org this is kinda a side hobby and doesn't pay the bills, so it gets attention on a limited basis. That said, if you discover that running the `replace_meta_box()` method on an earlier priority or a later priority helps resolve the compatibility issue you are seeing... I'm definitely open to that. But maybe we'll get lucky and registering them with the same name as core will resolve it. 🤞 

Thanks for your help on this!

**Comment by remcokalf** [comment]
I think $no_term using the singular_name makes sense. Its the same als 'Add new [term]' in singular also, below it. 

I'll do some more testing today for the Tabify Edit Screen incompatibilities. Yesterday I had the idea to remove the metaboxes on the option screen with a sepate function in functions.php, was halfway there to a solution maybe. But strange things are interfering: transients being set and after saving posts, lost/changed tab settings.. maybe even opcache involved.

As with you, this is not paying my bills. But I am sometimes too stubborn to let it go, and need closure or a decision to entirely drop it. And I always learn something from solving things like this. The amount of time it takes though...

I'll let you know the outcome. Don't let this keep you from shipping the new version if you want.

**Comment by remcokalf** [comment]
I just found an elegant and very simple way to support Tabify Edit Screen plugin:

It has an action called 'tabify_add_meta_boxes' with which you can add/remove metaboxes to/from their option screen. That solves all issues. 
This action has one parameter $post_type, which is the posttype for which you want to add/remove metaboxes.

We only have to pass our 'replace_meta_box()' function to this action  in class-wordpress-radio-taxonomy.php:

`add_action( 'tabify_add_meta_boxes', array( $this, 'replace_meta_box' ) );`

And change the function slightly to be able to pass a $post_type as parameter for when it is called this way. Then in the function we need to check if this paremeter is set not. If not, we are on a normal post edit page and need to use the post_type from `get_current_screen()` as before.

The beginning of the function would then look like this:

```
public function replace_meta_box( $post_type ) { //posttype parameter passed by action tabify_add_meta_boxes

    $screen = get_current_screen();
    
    if(!isset($post_type)) { //if no external $post_type passed
        $post_type = $screen instanceof WP_Screen ? $screen->post_type : '';
    }
```
The rest can stay the same.

In my testing this works very good, the radio metaboxes are 'sticking' okay on the Tabify Edit Screen options page, and the are on the right tab on the post edit screens. And they are not showing up double or disappearing again after post save (which was the case).

I am only wonderin if add_action() needs to be in class-wordpress-radio-taxonomy.php, or in the main plugin file, then we need to pass the parameter through launch(). Or not in RB4T at all and left to the user to add it in functions.php if needed (and document it). But I don't know how/if that can be made to work (can you call a plugin function like that from functions.php? My knowledge ends here...).

**Comment by helgatheviking** [comment]
Hi, so I was just looking at this right now and it seems to be fine without any additional code. 

image

What is the need for the `tabify_add_meta_boxes` callback? 

NB: I ran into the same PHP notices you'd reported in the [.org forums.](https://wordpress.org/support/topic/notices-in-php7-4/)

**Comment by remcokalf** [comment]
It is not that commit, I have added that change.

Your screenshot shows that the metabox is on the last tab, which is default for metaboxes it does not know about.
Try putting it on the first tab, then do some saves of the post and the Tabify options screen and see what happens.

There are a few problems:
- on the options page the radio button metaboxes are showing up together with their removed version, or not at all, or on the last tab (they don't stick to another tab). This depends on if you have saved a post that the metabox is on or not.
If the taxonomy is supported by multiple post types, it is even worse, saving another posttype deletes all the settings again.
- What I figured out is that the cause is that transients are set, then read, both ways (options screen and post types). That is why the options screen even knows about the added/removed metaboxes, because it has no get_current_screen()->post_type and our whole replace_meta_box function never runs. 
However when you start saving the option screen, it is getting a mess. 
The solution is to pass the function like I wrote, to the options screen, so also there, the add/remove metaboxes is known. Then it goes right both ways.

**Comment by helgatheviking** [comment]
Ahh I see what you mean. I'm not usually keen to add compatibility support directly into the plugin, but may be able to make an exception here since it's not too big a deal to add the existing function to an additional/custom hook.

Give this a test:.. it's based on your suggestion or reusing `replace_meta_box()`
https://github.com/helgatheviking/Radio-Buttons-for-Taxonomies/tree/tabify-compat

**Comment by remcokalf** [comment]
Looks good, thanks! I see no issues. 

I totally understand what you say about compatibility support. It would be great if you'd include it though.
If it is any consolation, I see that even big plugins like Advanced Custom Fields offer compatibility with this same hook from Tabify Edit Screen:
https://github.com/AdvancedCustomFields/acf/blob/master/includes/third-party.php

Have a great weekend!

**Comment by helgatheviking** [comment]
I'd definitely prefer to have it in a 3rdparty/compat class, but I think I might just leave it as written in the branch for now. Let me get this all merged in and and pushed. Shipping on a friday... yolo

Thanks for all your help on this!

**Comment by remcokalf** [comment]
No problem, it was fun, learned some stuff. Just updated and everything is running great.
Happy it all is working now! Thanks!

**Comment by helgatheviking** [comment]
I probably never would've gotten around to this without your interest. This is what makes open source fun. 

High five!
highfive

**Comment by remcokalf** [comment]
Hi, sorry to open this up again, but I had a new insight, leading to a possible simplification in RB4T:

I was implementing the same 'tabify_add_meta_boxes' hook for some other places in my theme where metaboxes are added/removed. I noticed that for some, Tabify Edit Screen settings screen was already recognizing them without adding this hook, and for others not. 

The ones that worked used the add_meta_boxes action, and had a callback function with $post_type as parameter (just like we added to the replace_meta_boxes() function in RB4T. 

Then I noticed that in Tabify Edit Screen, where the do_action for 'tabify_add_meta_boxes' is implemented, it also has a do_action for add_meta_boxes. See lines 99-102 of Tabify Edit Screen > inc > settings_posttype.php:

Schermafbeelding 2021-03-23 om 11 35 56

So, Tabify is already trying to detect added/removed metaboxes throught the 'add_meta_boxes' hook, but it only works when $post_type is present as parameter in the callback function of it.

Long story short, in RB4T the 'tabify_add_meta_boxes' action can be left out I think. It even is double. The RB4T metaboxes are already detected in Tabify, as long as we keep having $post_type as parameter of replace_meta_boxes() in RB4T.

I tested a bit and I see no issues when I comment out the add_action( 'tabify_add_meta_boxes') line in RB4T.

**Comment by helgatheviking** [comment]
Whoops, this got lost in my inbox. so you think we can remove the `add_action( 'tabify_add_meta_boxes') ` line? Doesn't sound urgent... it hasn't caused any problems and as my grandma used to say... it's not eating anything! :)

**Comment by remcokalf** [comment]
Yes, exactly. (As long as the replace_meta_box function runs on the 'add_meta_boxes' and it has a $post_type parameter, as is the situation now.)

It's not urgent. 
It not eating anything except some resources and loading time I guess. But as this rewrite was mostly about optimizing the plugin, it would be worth it to remove it. The action is double now, so the replace_meta_boxes() function called from Tabify runs double the amount of times needed (on every metabox, on the Tabify Edit Sceen settings window). And it is adding+removing all radio metaboxes twice.
I put up a counter on the replace_meta_boxes() function. In my case it runs 180 times with the double action and 90 times without.

**Comment by helgatheviking** [comment]
A PR might help speed it along if you are so inclined. :) I'm pretty deep in a major update to another plugin at the moment.

**Comment by helgatheviking** [comment]
Released! Thanks again.

**Comment by remcokalf** [comment]
Great, thanks!
